Modify rule S1046: LaYC format (#2619)
SonarSource/rspec: rules/S1046/cfamily/rule.adoc

== Why is this an issue?

> The catch-all handler, written `catch(...)` in C{plus}{plus}, or `@catch(...)` in Objective-C, catches every type of exception. If there is another catch statement for a specific exception after the catch-all handler, it will not be executed because the catch-all handler will already have handled the exception.

Reviewer: The previous Asciidoc version was also incorrect: the `...` needs to be escaped with `` `+ bla +` `` or it is transformed to the Unicode equivalent … .

```suggestion
The catch-all handler, written `+catch(...)+` in {cpp}, or `+@catch(...)+` in Objective-C, catches every type of exception. If there is another catch statement for a specific exception after the catch-all handler, it will not be executed because the catch-all handler will already have handled the exception.
```

BTW, Grammarly suggests splitting the second sentence for clarity:
```suggestion
The catch-all handler, written `+catch(...)+` in {cpp}, or `+@catch(...)+` in Objective-C, catches every type of exception. Suppose there is another catch statement for a specific exception after the catch-all handler. In that case, it will not be executed because the catch-all handler will already have handled the exception.
```

== How to fix it

=== Code examples

> The following C{plus}{plus} examples also work in Objective-C by replacing `try` and `catch` with their equivalents `@try` and `@catch`.

Reviewer: To be extremely pedantic, the exception type should also be changed (`std::exception const &e` --> `NSException *e`). It's probably not worth going into so much detail here though. I'd just change the sentence to be accurate, maybe something like this:

```suggestion
The following {cpp} example is very similar in Objective-C: the `try` and `catch` equivalents are `@try` and `@catch`, respectively.
```

Author: Modified

==== Noncompliant code example

[source,cpp]
----
void f() {
  try {
    // ...
  } catch (...) {
    // Handle all exception types
  } catch (std::exception const &e) { // Noncompliant: it will never be called
  }
}
----

==== Compliant solution

[source,cpp]
----
void f() {
  try {
    // ...
  } catch (std::exception const &e) {
    // Handle standard exceptions
  } catch (...) { // Compliant: handle all other exception types
  }
}
----

== Resources

=== Documentation

> * C{plus}{plus} reference - https://en.cppreference.com/w/cpp/language/try_catch[try-block]

Reviewer: I'm not 100% convinced it's necessary but we could link to https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Exceptions/Tasks/HandlingExceptions.html for completeness.

WDYT?

Author: "library/archive" :D
I will add it for the unlucky users who deal with Obj-C

> * C{plus}{plus} reference - https://en.cppreference.com/w/cpp/language/try_catch[try-block]

Reviewer: ```suggestion
* {cpp} reference - https://en.cppreference.com/w/cpp/language/try_catch[try-block]
```

For consistency with the idiomatic way of spelling C++ in Asciidoc.

Author: Oh indeed! I don't know why I forgot about it... Thanks!
